File: backend/src/shared_lib/entities/assessment.py
from enum import StrEnum
from typing import Any, TypedDict

from pydantic import BaseModel

from entities.question import Pillar

AssessmentID = str


class Steps(StrEnum):
    SCANNING_STARTED = "SCANNING_STARTED"
    PREPARING_PROMPTS = "PREPARING_PROMPTS"
    INVOKING_LLM = "INVOKING_LLM"
    FINISHED = "FINISHED"
    ERRORED = "ERRORED"


class AssessmentData(TypedDict):
    regions: dict[str, int]
    resource_types: dict[str, int]
    severities: dict[str, int]
    findings: int


class Assessment(BaseModel):
    id: AssessmentID
    name: str
    regions: list[str] = []
    role_arn: str
    workflows: list[str] = []
    step: Steps
    execution_arn: str | None = None
    created_at: str
    raw_graph_datas: dict[str, AssessmentData] | None = None
    graph_datas: AssessmentData | None = None
    error: dict[str, Any] | None = None
    question_version: str | None = None
    findings: dict[str, Pillar] | None = None


class AssessmentDto(BaseModel):
    name: str | None = None
    role_arn: str | None = None
    created_at: str | None = None
    step: Steps | None = None
    raw_graph_datas: dict[str, AssessmentData] | None = None
    graph_datas: AssessmentData | None = None
    error: dict[str, Any] | None = None
    question_version: str | None = None
    findings: dict[str, Pillar] | None = None

File: libs/backend/models/src/lib/Assessment/Assessment.ts
import { type BillingInformation } from '../BillingInformation';
import type { Pillar } from '../Pillar';

export interface Assessment {
  createdAt: Date;
  createdBy: string;
  executionArn: string;
  pillars?: Pillar[];
  id: string;
  name: string;
  organization: string;
  questionVersion?: string;
  regions: string[];
  exportRegion?: string;
  roleArn: string;
  workflows: string[];
  finishedAt?: Date;
  error?: AssessmentError;
  fileExports?: AssessmentFileExport[];
  wafrWorkloadArn?: string;
  opportunityId?: string;
  opportunityCreatedAt?: Date;
  billingInformation?: BillingInformation;
  folder?: string;
}

export enum AssessmentStep {
  SCANNING_STARTED = 'SCANNING_STARTED',
  PREPARING_ASSOCIATIONS = 'PREPARING_ASSOCIATIONS',
  ASSOCIATING_FINDINGS = 'ASSOCIATING_FINDINGS',
  FINISHED = 'FINISHED',
  ERRORED = 'ERRORED',
}

export interface AssessmentError {
  cause: string;
  error: string;
}

export interface AssessmentBody {
  name?: string;
  error?: AssessmentError;
  finishedAt?: Date;
  questionVersion?: string;
  exportRegion?: string;
  wafrWorkloadArn?: string;
  opportunityId?: string;
  executionArn?: string;
  opportunityCreatedAt?: Date;
  folder?: string | null;
}

export enum AssessmentFileExportType {
  PDF = 'pdf',
}

export enum AssessmentFileExportStatus {
  NOT_STARTED = 'NOT_STARTED',
  IN_PROGRESS = 'IN_PROGRESS',
  COMPLETED = 'COMPLETED',
  ERRORED = 'ERRORED',
}

export interface AssessmentFileExport {
  id: string;
  status: AssessmentFileExportStatus;
  type: AssessmentFileExportType;
  error?: string;
  versionName: string;
  objectKey?: string;
  createdAt: Date;
}
